### ISSUE TITLE
TF1 and Keras fine-tuning results differ significantly

### ISSUE THREAD

**Issue by himanshurawlani** [google/automl#848]
I am finetuning EfficientDet D0 model (pretrained on COCO dataset) on PASCAL VOC dataset. The model trains successfully using `main.py` (AP [IoU=0.50] = 0.757) however, the same model doesn't train using `keras/train.py` (AP [IoU=0.50] = 0.005).

I use the following commands to create training and validation tfrecords:

```
# Create tfrecords for training
PYTHONPATH=".:$PYTHONPATH" python dataset/create_pascal_tfrecord.py \
    --data_dir=VOCdevkit --year=VOC2007 --output_path=tfrecords/pascal_train --set=train

# Create tfrecords for validation
PYTHONPATH=".:$PYTHONPATH" python dataset/create_pascal_tfrecord.py \
    --data_dir=VOCdevkit --year=VOC2007 --output_path=tfrecords/pascal_val --set=val
```
I train EfficientDet D0 model using the following command for main.py:

```
python main.py --mode=train_and_eval \
    --training_file_pattern=tfrecord/pascal_train*.tfrecord \
    --validation_file_pattern=tfrecord/pascal_val*.tfrecord \
    --model_name=efficientdet-d0 \
    --model_dir=efficientdet-d0-finetune-pascal  \
    --ckpt=efficientdet-d0  \
    --train_batch_size=32 \
    --eval_batch_size=32 --eval_samples=2510 \
    --num_examples_per_epoch=2501 --num_epochs=50  \
    --hparams=voc_config.yaml \
    --strategy=gpus
```
And I get the following results:

```
AP = 0.5191394, AP50 = 0.7567504
det_loss = 0.4822153, cls_loss = 0.31231615, box_loss = 0.003397982, loss: 0.57546026
```

However, when I train EfficientDet D0 model using the following command for keras/train.py:

```
PYTHONPATH=".:$PYTHONPATH" python keras/train.py \
    --training_file_pattern=tfrecords/pascal_train*.tfrecord \
    --validation_file_pattern=tfrecords/pascal_val*.tfrecord \
    --model_name=efficientdet-d0 \
    --model_dir=efficientdet-d0-finetune-pascal \
    --pretrained_ckpt=efficientdet-d0 \
    --batch_size=32 --eval_samples=2510 \
    --num_examples_per_epoch=2501 --num_epochs=50 \
    --hparams=voc_config.yaml \
    --strategy=gpus
```

I get the following results:

```
AP = 0.0, AP50 = 0.005
det_loss: 1.4360 - cls_loss: 0.9178 - box_loss: 0.0104 - loss: 1.4481
```

I am using the same tfrecords and config file for both the scripts (main.py and keras/train.py). Below is the config file:

```
num_classes: 21
var_freeze_expr: '(efficientnet|fpn_cells|resample_p6)'
label_map: {1: aeroplane, 2: bicycle, 3: bird, 4: boat, 5: bottle, 6: bus, 7: car, 8: cat, 9: chair, 10: cow, 11: diningtable, 12: dog, 13: horse, 14: motorbike, 15: person, 16: pottedplant, 17: sheep, 18: sofa, 19: train, 20: tvmonitor}
```

I am expecting the loss values for both the scripts to be in the same ballpark. I am not able to debug the issue, any help would be appreciated.

**Comment by jonkoi** [comment]
I think he Keras version is not working correctly now.

In the function keras/restore_ckpt.py, `var_dict` is used as a dictionary of variables to be assigned values from the checkpoint. But if I print it out, it is empty.

So I think you trained the keras model from scratch. Together with the freezing, the performance is much worse.

**Comment by himanshurawlani** [comment]
I tried evaluating PASCAL VOC fine-tuned model (trained using main.py) using keras/eval.py and I got the same results as that of main.py. Also, I didn't get any warnings when loading the fine-tuned model (training using main.py). However, when I load model trained using keras/train.py, I get multiple warnings of the format:

`WARNING:tensorflow:Unresolved object in checkpoint: (root).optimizer's state 'average' for (root).fpn_cells.cells.0.fnodes.5.resample_layers.0.bn.beta`

**Comment by jonkoi** [comment]
I get warnings like yours, too. But I have no ideas what they are about.

keras/eval.py seems to use different way to restore ckpt than keras/train.py (`model.load_weights` instead of the custom `keras_utils.restore_ckpt`). So I still suspect that the training script does not restore the checkpoint correctly.

**Comment by himanshurawlani** [comment]
Yes, looks like `keras/train.py` script doesn't restore the checkpoint correctly. I downloaded `.h5` pretrained COCO checkpoint and passed it directly to `model.load_weights()` function in `keras/train.py`:

```
model.build((None, *config.image_size, 3))
if FLAGS.pretrained_ckpt:
    # ckpt_path = tf.train.latest_checkpoint(FLAGS.pretrained_ckpt)
    # util_keras.restore_ckpt(model, ckpt_path, params['moving_average_decay'])
    model.load_weights(FLAGS.pretrained_ckpt, by_name=True, skip_mismatch=True)
```
After training the model for 50 epochs, I got the following results:

```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.404
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.668
```
Does this work for you too? @jonkoi

**Comment by fsx950223** [comment]
Fixed in https://github.com/google/automl/commit/aa6480fe7e07bd99030e56b7f05c75e5291db357
